api/api-private-cluster-nodes.go

```diff
@@ -0,0 +1,225 @@
+// This file is part of MinIO Kubernetes Cloud
+// Copyright (c) 2019 MinIO, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strconv"
+	"strings"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/minio/m3/cluster"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/minio/m3/api/stubs"
+)
+
+// ClusterNodesAdd rpc to add a new node to the cluster
+func (ps *privateServer) ClusterNodesAdd(ctx context.Context, in *pb.NodeAddRequest) (*pb.NodeAddResponse, error) {
+	if in.Name == "" {
+		return nil, status.New(codes.InvalidArgument, "A node name is needed").Err()
+	}
+
+	if in.K8SLabel == "" {
+		return nil, status.New(codes.InvalidArgument, "A kubernetes label is needed").Err()
+	}
+
+	// validate volumes string
+	if in.Volumes == "" {
+		return nil, status.New(codes.InvalidArgument, "A list of volumes on the node is needed").Err()
+	}
+
+	var volumes []string
+
+	if strings.Contains(in.Volumes, "...") {
+		var re = regexp.MustCompile(`^.*?({([0-9]+)\.\.\.([0-9]+)}).*?$`)
+		if !re.MatchString(in.Volumes) {
+			return nil, status.New(codes.InvalidArgument, "Invalid descriptor of volumes mount points").Err()
+		}
+		rs := re.FindStringSubmatch(in.Volumes)
+		// parse nums, regex should have validated valid integers
+		fromNum, _ := strconv.Atoi(rs[2])
+		toNum, _ := strconv.Atoi(rs[3])
+		// if we received /mnt/disk{1...4} we will iterate 4 times, and replace {1...4} for the corresponding number
+		for i := fromNum; i <= toNum; i++ {
+			volumes = append(volumes, strings.Replace(in.Volumes, rs[1], fmt.Sprintf("%d", i), 1))
+		}
+	} else {
+		// attempt to validate single volume mount path
+		randomNodeID := uuid.NewV4()
+		if _, err := cluster.NewVolume(&randomNodeID, in.Volumes); err != nil {
+			return nil, status.New(codes.InvalidArgument, "Volume mount path is not valid").Err()
```

Comment:
we are using a specific error related to all errors that can happen during `NewVolume()` we should probably use a more generic message

Reply:
This is merely to use NewVolume for validation, it needs some nodeID so I'm randomly generating one.

Reply:
A more clear error message was added